### PR TITLE
Periodic changes

### PR DESCRIPTION
**Pull request by andrewtarzia**
In response to #182 .

Adds Writers (PdbWriter() for now) and TopologyGraph.get_periodic_cell().

Needs tests and splitting into their own PRs.

### CODE REVIEW

**Comment by lukasturcani** [comment]
This is great, tyvm!

Feel free to bump me on this if I don't get back to you this week.

File: src/stk/molecular/writers/mdl_mol.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+            'M  V30 BEGIN BOND\n'
+            f'{bond_block}'
+            'M  V30 END BOND\n'
+            'M  V30 END CTAB\n'
+            'M  END\n'
+            '\n'
+            '$$$$\n'
+        )
+
+    def write_string(self, molecule, atom_ids=None):
```
Comment (lukasturcani) [inline src/stk/molecular/writers/mdl_mol.py:91]: Maybe call this `to_string` ?

File: src/stk/molecular/topology_graphs/cof/periodic_hexagonal.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    def get_building_blocks(self):
+        yield from self._internal.get_building_blocks()
+
+    def get_num_building_block(self, building_block):
+        return self._internal.get_num_building_block(building_block)
+
+    def clone(self):
+        return self._internal.clone()
+
+    def get_periodic_cell(self):
```
Comment (lukasturcani) [inline src/stk/molecular/topology_graphs/cof/periodic_hexagonal.py:128]: What do you think about this being `get_periodic_info` and returning a `PeriodicInfo` instance?

File: src/stk/molecular/topology_graphs/cof/__init__.py [hunk 3]
```diff
@@ -4,3 +4,4 @@
 from .linkerless_honeycomb import *  # noqa
 from .square import *  # noqa
 from .cof import *  # noqa
+from .periodic_hexagonal import *  # noqa
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/cof/__init__.py:7]: Handled.

File: src/stk/molecular/topology_graphs/topology_graph/topology_graph/topology_graph.py [hunk 4]
```diff
@@ -407,6 +407,24 @@ def _get_construction_state(self):
             )
         )
 
+    def get_periodic_info(self):
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/topology_graph/topology_graph/topology_graph.py]: Done.

File: src/stk/molecular/topology_graphs/cof/cof.py [hunk 5]
```diff
@@ -589,6 +615,35 @@ def _get_scale(self, building_block_vertices):
             for bb in building_block_vertices
         )
 
+    def get_periodic_cell(self):
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/cof/cof.py]: Done

File: src/stk/molecular/writers/mdl_mol.py (hunk 1 above)
Comment (andrewtarzia) [reply src/stk/molecular/writers/mdl_mol.py:91]: Done.

File: src/stk/molecular/topology_graphs/cof/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/cof/__init__.py:7]: Done.

File: src/stk/molecular/topology_graphs/cof/cof.py [hunk 6]
```diff
@@ -77,6 +88,23 @@ class Cof(TopologyGraph):
...
+
+    .. code-block:: python
+
+        topology_graph = stk.cof.Honeycomb(
+            building_blocks=(bb1, bb2),
+            lattice_size=(3, 3, 1),
+            periodic=True,
+        )
+        cof = stk.ConstructedMolecule(topology_graph)
+        cell_matrix = topology.get_periodic_cell()
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/cof/cof.py:106]: Done.

File: src/stk/molecular/topology_graphs/cof/periodic_hexagonal.py (hunk 2 above)
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/cof/periodic_hexagonal.py:128]: Done after discussion.

File: src/stk/molecular/writers/turbomole.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+"""
+
+
+class TurbomoleWriter:
+    """
+    A writer class for ``Turbomole`` files.
+
+    Examples
+    --------
+    *Writing to File with Unit Cell*
```
Comment (lukasturcani) [inline src/stk/molecular/writers/turbomole.py:19]: a File?

File: src/stk/molecular/writers/turbomole.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+    This writer can write to file for visualisation with the unit cell
```
Comment (lukasturcani) [inline src/stk/molecular/writers/turbomole.py:21]: a file?

File: src/stk/molecular/writers/turbomole.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+        return content
+
+    def to_string(
+        self,
+        molecule,
+        atom_ids=None,
+        periodic_info=None
+    ):
+        """
+        Write `molecule` to ``Turbomole`` file format as string.
```
Comment (lukasturcani) [inline src/stk/molecular/writers/turbomole.py:86]:
```
Get a ``Turbomole`` file format string of `molecule`.
```

File: src/stk/molecular/writers/turbomole.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+        molecule : :class:`.Molecule`
+            Molecule to write to ``Turbomole`` format.
+
+        atom_ids : :class:`iterable` of :class:`int`
+            The atom ids of atoms to write. Can be a single
+            :class:`int`, if a single atom is to be used, or ``None``,
+            if all atoms are to be used.
+
+        periodic_info : :class:`.PeriodicInfo`
+            Information about periodic cell.
```
Comment (lukasturcani) [inline src/stk/molecular/writers/turbomole.py:99]: about the

File: src/stk/molecular/writers/turbomole.py [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+
+        Returns
+        -------
+        string :class:`string`
```
Comment (lukasturcani) [inline src/stk/molecular/writers/turbomole.py:103]:
```
:class:`str`
```

File: src/stk/molecular/writers/turbomole.py [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+            The string containing all information for a `.coord` file.
```
Comment (lukasturcani) [inline src/stk/molecular/writers/turbomole.py:104]:
```
The content of a `.coord` file.
```

File: src/stk/molecular/writers/turbomole.py [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+        path : :class:`str`
+            The full path to the file being written.
+
+        atom_ids : :class:`iterable` of :class:`int`
+            The atom ids of atoms to write. Can be a single
+            :class:`int`, if a single atom is to be used, or ``None``,
+            if all atoms are to be used.
+
+        periodic_info : :class:`.PeriodicInfo`
+            Information about periodic cell.
```
Comment (lukasturcani) [inline src/stk/molecular/writers/turbomole.py:134]: about the

File: src/stk/molecular/writers/xyz.py [hunk 14]
```diff
@@ -0,0 +1,105 @@
+"""
+XyzWriter
```
Comment (lukasturcani) [inline src/stk/molecular/writers/xyz.py:2]: XYZ Writer (it's a title)

File: src/stk/molecular/writers/xyz.py [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+=========
+
+.. toctree::
+    :maxdepth: 2
+
+    XyzWriter <stk.molecular.writers.xyz>
```
Comment (lukasturcani) [inline src/stk/molecular/writers/xyz.py:8]: XYZ Writer (also title)

File: src/stk/molecular/writers/xyz.py [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+
+"""
+
+
+class XyzWriter:
+    """
+    A writer class for `.xyz` files.
```
Comment (lukasturcani) [inline src/stk/molecular/writers/xyz.py:15]:
```
``.xyz``
```

File: src/stk/molecular/writers/xyz.py [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+
+    Examples
+    --------
+    *Writing to File*
```
Comment (lukasturcani) [inline src/stk/molecular/writers/xyz.py:19]: to a

File: src/stk/molecular/writers/xyz.py [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+            symbol = atom.__class__.__name__
+            content.append(f'{symbol} {x:f} {y:f} {z:f}\n')
+        # Set first line to the atom_count.
+        content[0] = f'{i}\n\n'
+
+        return content
+
+    def to_string(self, molecule, atom_ids=None):
+        """
+        Write `molecule` to `.xyz` format as string.
```
Comment (lukasturcani) [inline src/stk/molecular/writers/xyz.py:55]:
```
Get the ``.xyz`` string of  `molecule`.
```

File: src/stk/molecular/writers/xyz.py [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+            Molecule to write to `.xyz` format.
+
+        atom_ids : :class:`iterable` of :class:`int`
+            The atom ids of atoms to write. Can be a single
+            :class:`int`, if a single atom is to be used, or ``None``,
+            if all atoms are to be used.
+
+        Returns
+        -------
+        content : :class:`str`
```
Comment (lukasturcani) [inline src/stk/molecular/writers/xyz.py:69]:
```
:class:`str`
```

File: src/stk/molecular/writers/xyz.py [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+            String in `.xyz` file format.
```
Comment (lukasturcani) [inline src/stk/molecular/writers/xyz.py:70]:
```
``.xyz``

```

File: src/stk/molecular/writers/xyz.py [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+        """
+        Write `molecule` to `.xyz` file format.
+
+        Parameters
+        ----------
+        molecule : :class:`.Molecule`
+            Molecule to write to `.xyz` format.
+
+        path : :class:`str`
+            The full path to the file being written..
```
Comment (lukasturcani) [inline src/stk/molecular/writers/xyz.py:88]: double full stop here.

File: src/stk/molecular/writers/xyz.py [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+        return ''.join(content)
+
+    def write(self, molecule, path, atom_ids=None):
+        """
+        Write `molecule` to `.xyz` file format.
+
+        Parameters
+        ----------
+        molecule : :class:`.Molecule`
+            Molecule to write to `.xyz` format.
```
Comment (lukasturcani) [inline src/stk/molecular/writers/xyz.py:85]:
```
``.xyz``
```

File: src/stk/molecular/molecules/periodic_info.py [hunk 23]
```diff
@@ -0,0 +1,124 @@
+"""
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/molecules/periodic_info.py:1]: Done.

File: src/stk/molecular/molecules/periodic_info.py [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+    """
+
+    def __init__(self, cell_matrix):
+        """
+        Initialize a :class:`PeriodicInfo` instance.
+
+        Parameters
+        ----------
+        cell_matrix : :class:`tuple` of :class:`np.array`
+            Tuple of cell lattice vectors (shape: (3,)) in Angstrom.
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/molecules/periodic_info.py:29]: This is handled by splitting the input into the x_*, y_* and z_vector, which are the three components of the matrix and are each np arrays of size 3

File: src/stk/molecular/molecules/periodic_info.py [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+        return clone
+
+    def get_cell_matrix(self):
+        return self._cell_matrix
+
+    def _with_cell_matrix(self, cell_matrix):
+        self._cell_matrix = cell_matrix
+        return self
+
+    def with_cell_matrix(self, cell_matrix):
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/molecules/periodic_info.py:60]: Done.

File: src/stk/molecular/molecules/periodic_info.py [hunk 26]
```diff
@@ -0,0 +1,124 @@
...
+        """
+        Return a clone with cell_matrix.
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/molecules/periodic_info.py:62]: Done

File: src/stk/molecular/molecules/periodic_info.py [hunk 27]
```diff
@@ -0,0 +1,124 @@
...
+
+        Returns
+        -------
+        :class:`.PeriodicInfo`
+            The clone. Has the same cell as the original.
+
+        """
+        return self.clone()._with_cell_matrix(cell_matrix)
+
+    def get_lengths_and_angles(self):
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/molecules/periodic_info.py:77]: Agreed, and I have made those changes. Although my naming matches what is expected given the standard naming conventions for unit cells.

File: src/stk/molecular/molecules/periodic_info.py [hunk 28]
```diff
@@ -0,0 +1,124 @@
...
+
+    """
+
+    def __init__(self, cell_matrix):
+        """
+        Initialize a :class:`PeriodicInfo` instance.
+
+        Parameters
+        ----------
+        cell_matrix : :class:`tuple` of :class:`np.array`
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/molecules/periodic_info.py:28]: Done

File: src/stk/molecular/writers/mdl_mol.py [hunk 29]
```diff
@@ -0,0 +1,139 @@
+"""
+MolWriter
+=========
+
+.. toctree::
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/writers/mdl_mol.py:5]: Done.

File: src/stk/molecular/writers/mdl_mol.py [hunk 30]
```diff
@@ -0,0 +1,139 @@
...
+"""
+
+
+class MolWriter:
+    """
+    A writer class for V3000 ``.mol`` files.
+
+    Examples
+    --------
+    *Writing to File*
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/writers/mdl_mol.py:19]: Made the change. I think both work and I agree that 'a file' sounds more natural.

File: src/stk/molecular/writers/pdb.py [hunk 31]
```diff
@@ -0,0 +1,190 @@
...
+        if atom_ids is None:
+            atom_ids = range(molecule.get_num_atoms())
+        elif isinstance(atom_ids, int):
+            atom_ids = (atom_ids, )
+
+        content = []
+        if periodic_info is not None:
+            # Input unit cell information.
+            lengths_and_angles = periodic_info.get_lengths_and_angles()
+            a, b, c, alpha, beta, gamma = lengths_and_angles
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/writers/pdb.py:56]: Agreed!

File: src/stk/molecular/writers/pdb.py [hunk 32]
```diff
@@ -0,0 +1,190 @@
...
+                f'{element:>2}{charge:>2}\n'
+            )
+
+        conect = 'CONECT'
+        for bond in molecule.get_bonds():
+            a1 = bond.get_atom1().get_id()
+            a2 = bond.get_atom2().get_id()
+            # # Do not form periodic bonds in a PDB.
+            # if periodic_cell is not None and bond.is_periodic():
+            #     continue
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/writers/pdb.py:103]: I kept this here to check if I wanted the code to do this or not. I have decided to remove the commented code. I.e. bonds across periodic boundaries should be written.

File: src/stk/molecular/writers/pdb.py [hunk 33]
```diff
@@ -0,0 +1,190 @@
...
+            atom_ids=atom_ids,
+            periodic_info=periodic_info,
+        )
+
+        return ''.join(content)
+
+    def write(
+        self,
+        molecule,
+        path=None,
```
[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/stk/molecular/writers/pdb.py:153]: Mistake from old code.

**Review by lukasturcani — APPROVED** [review]

**Pull request merged**